src/shared/tests.c

```diff
@@ -36,33 +36,37 @@ char* setup_fake_runtime_dir(void) {
         return p;
 }
 
-const char* get_testdata_dir(void) {
+const char* get_testdata_dir(const char *suffix) {
         const char *env;
-        _cleanup_free_ char *exedir = NULL;
         /* convenience: caller does not need to free result */
         static char testdir[PATH_MAX];
 
         /* if the env var is set, use that */
         env = getenv("SYSTEMD_TEST_DATA");
         if (env) {
-                if (access(env, F_OK) >= 0)
-                        return env;
-
-                fputs("ERROR: $SYSTEMD_TEST_DATA directory does not exist\n", stderr);
-                exit(1);
+                if (access(env, F_OK) < 0) {
+                        fputs("ERROR: $SYSTEMD_TEST_DATA directory does not exist\n", stderr);
+                        exit(1);
+                }
+                strncpy(testdir, env, sizeof(testdir) - 1);
```

Comment:
I did after the strncpy() below; but indeed this wouldn't suffice as not terminating it here would confuse the second strncpy()'s strlen() call. Sorry about that. (Then again, there are probably easier ways to do harm to yourself than trying to inject bogus SYSTEMD_TEST_DATA values, but let's be correct here indeed.)

Also, the termination shouldn't be necessary in the first place as static variables (like global ones) should get nulled at program start. But efficiency doesn't matter here. I pushed again to move the null termination further up.